pkg/recovery/apiserver.go

```diff
@@ -0,0 +1,438 @@
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"path"
+	"text/template"
+	"time"
+
+	"github.com/ghodss/yaml"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/runtime/serializer"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/apiserver/pkg/authentication/user"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	clientcmdapiv1 "k8s.io/client-go/tools/clientcmd/api/v1"
+	"k8s.io/klog"
+
+	"github.com/openshift/library-go/pkg/crypto"
+
+	"github.com/openshift/cluster-kube-apiserver-operator/pkg/operator/v311_00_assets"
+)
+
+const (
+	KubeApiserverStaticPodFileName = "kube-apiserver-pod.yaml"
+	RecoveryPodFileName            = "recovery-kube-apiserver-pod.yaml"
+	RecoveryPodNamespace           = "kube-system"
+	RecoveryCofigFileName          = "config.yaml"
+	AdminKubeconfigFileName        = "admin.kubeconfig"
+
+	RecoveryPodAsset    = "v3.11.0/kube-apiserver/recovery-pod.yaml"
+	RecoveryConfigAsset = "v3.11.0/kube-apiserver/recovery-config.yaml"
+)
+
+var (
+	Scheme = runtime.NewScheme()
+	Codecs = serializer.NewCodecFactory(Scheme)
+)
+
+func init() {
+	metav1.AddToGroupVersion(Scheme, schema.GroupVersion{Version: "v1"})
+	utilruntime.Must(corev1.AddToScheme(Scheme))
+}
+
+type Apiserver struct {
+	KubeApiserverImage string
+	PodManifestDir     string
+	ResourceDirPath    string
+	RecoveryDirPath    string
+
+	kubeApiserverStaticPod *corev1.Pod
+	restConfig             *rest.Config
+	kubeClientSet          *kubernetes.Clientset
+}
+
+func (s *Apiserver) GetCurrentApiserverManifest() (*corev1.Pod, error) {
+	if s.kubeApiserverStaticPod != nil {
+		return s.kubeApiserverStaticPod, nil
+	}
+
+	kubeApiserverManifestPath := path.Join(s.PodManifestDir, KubeApiserverStaticPodFileName)
+	f, err := os.OpenFile(kubeApiserverManifestPath, os.O_RDONLY, 0644)
+	if err != nil {
+		return nil, fmt.Errorf("failed to open file %q: %v", kubeApiserverManifestPath, err)
+	}
+	defer f.Close()
+
+	buf := bytes.NewBuffer(nil)
+	_, err = io.Copy(buf, f)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read file %q: %v", kubeApiserverManifestPath, err)
+	}
+
+	obj, err := runtime.Decode(Codecs.UniversalDecoder(corev1.SchemeGroupVersion), buf.Bytes())
+	if err != nil {
+		return nil, fmt.Errorf("failed to decode file %q: %v", kubeApiserverManifestPath, err)
+	}
+
+	// TODO: support conversions if the object is in different but convertible version
+	kubeApiserverStaticPod, ok := obj.(*corev1.Pod)
+	if !ok {
+		return nil, fmt.Errorf("unsupported type: kubeApiserverStaticPodManifest is not type *corev1.Pod but %T", obj)
+	}
+
+	s.kubeApiserverStaticPod = kubeApiserverStaticPod
+
+	return s.kubeApiserverStaticPod, nil
+}
+
+func (s *Apiserver) GetResourceDirPath() (string, error) {
+	if s.ResourceDirPath != "" {
+		return s.ResourceDirPath, nil
```

Comment:
is this really time critical that logic like this makes sense?

Reply:
some caller don't need it some don't, better to fail only reading stuff you need

Reply:
> some caller don't need it some don't, better to fail only reading stuff you need

I agree with sttts.  Caching is harder for me to think about as I read than non-caching.

Reply:
Was not intended as caching initially but as deferred loading; the other PR adds another caller creating some distinct paths. removed most of it because it was not needed anymore (I guess I'll find that out when I do the rebase :)).